File: data_structures/stack.py
class Node:
    def __init__(self, val):
      self.value = val
      self.next = None


class Stack:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    def push(self, val):
        node = Node(val)
        if not self.first:
            self.first = node
            self.last = node
        else:
            node.next = self.first
            self.first = node

        self.size += 1

        return self.size

    def pop(self):
        if not self.first: return None

        node = self.first

        if self.size == 1:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next

        node.next = None
        self.size -= 1

        return node.value


    def __repr__(self):
        items = []
        current = self.first
        while current:
            items.append(str(current.value))
            current = current.next
        return f"Stack (top → bottom): [{' -> '.join(items)}]"

stack = Stack()

print(stack.push(3))
print(stack.push(2))
print(stack.push(5))
print(stack.push(7))

print(stack)

print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())

print(stack)

File: data_structures/stack.ts
class StackNode {
  value: any
  next: (StackNode | null)

  constructor(value: any) {
    this.value = value
    this.next = null
  }
}

class Stack {
  first: (StackNode | null);
  last: (StackNode | null);
  size: number;

  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }

  push(val: any) {
    const new_node = new StackNode(val)

    if(!this.last) {
      this.last = new_node;
    } else {
      new_node.next = this.first;
    }

    this.first = new_node;
    this.size++;

    return this.size;
  }

  pop() {
    if(!this.first) return;

    const old_head = this.first;

    if(this.size === 1) {
      this.first = null;
      this.last = null;
    } else {
      this.first = this.first.next;
    }

    this.size--;

    return old_head.value;
  }

  print() {
    const items = [];
    let curr = this.first;

    while(curr) {
      items.push(curr.value);
      curr = curr.next;
    }

    console.log(items)
  }
}

const stack = new Stack();

console.log(stack.push(3));
console.log(stack.push(2));
console.log(stack.push(5));
console.log(stack.push(7));

console.log(stack);

console.log(stack.pop());
console.log(stack.pop());
console.log(stack.pop());
console.log(stack.pop());
console.log(stack.pop());
console.log(stack.pop());

console.log(stack);
